modules/pam_faillock/main.c

```diff
@@ -51,29 +51,37 @@
 #define AUDIT_NO_ID     ((unsigned int) -1)
 #endif
 
+#include "pam_inline.h"
 #include "faillock.h"
-
-struct options {
-	unsigned int reset;
-	const char *dir;
-	const char *user;
-	const char *progname;
-};
+#include "faillock_config.h"
 
 static int
 args_parse(int argc, char **argv, struct options *opts)
 {
 	int i;
+	int rv;
+	const char *conf = default_faillock_conf;
+
 	memset(opts, 0, sizeof(*opts));
 
-	opts->dir = FAILLOCK_DEFAULT_TALLYDIR;
+	opts->dir = strdup(FAILLOCK_DEFAULT_TALLYDIR);
 	opts->progname = argv[0];
 
 	for (i = 1; i < argc; ++i) {
-		if (strcmp(argv[i], "--dir") == 0) {
+		if (strcmp(argv[i], "--conf") == 0) {
 			++i;
 			if (i >= argc || strlen(argv[i]) == 0) {
-				fprintf(stderr, "%s: No directory supplied.\n", argv[0]);
+				fprintf(stderr, "%s: No configuration directory supplied.\n",
+						argv[0]);
+				return -1;
+			}
+			conf = pam_str_skip_prefix(argv[i], "--conf=");
+		}
+		else if (strcmp(argv[i], "--dir") == 0) {
+			++i;
+			if (i >= argc || strlen(argv[i]) == 0) {
+				fprintf(stderr, "%s: No records directory supplied.\n",
+						argv[0]);
 				return -1;
 			}
 			opts->dir = argv[i];
```

Comment:
The way it is implemented the value from configuration file if it is set there always overrides the --dir option.

Reply:
This part of the code loads all the command line arguments in variables, but it doesn't set any priority for them. The coding lines that sets the priority for the tally directory are located in [faillock_config.c 169-178](https://github.com/linux-pam/linux-pam/pull/457/files#diff-e188da21b29b52dc3cac68918c0cc91d437eb192ccf9fba04cf19c7dbf9c72a0R169-R178).

Reply:
Hmm, it still means that `--dir /var/run/faillock` will be overridden.

Reply:
And also you need to use `opts->dir = strdup(argv[i]);` here.

Reply:
I recommend instead of doing the strcmp in faillock_config, just save the dir option value here and override it in the opts after read_config_file() is called if it is non-NULL.

Reply:
Based on your comment I have changed this part of the logic.

I also had to change other things to accommodate everything so I added a new commit. I did this to get a clear view of what changed. If you agree with those changes I'll squash the commit.

Reply:
I just squashed the commits and improved the commit message wording.